In [4]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Import Plotting Libararies
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

# Model Evaluation Libraries
from sklearn.metrics import classification_report, confusion_matrix

In [5]:
#expanding the dispay of text sms column
pd.set_option('display.max_colwidth', -1)

dataset = pd.read_csv("sarcasm_dataset.csv")
dataset.head()

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  


Unnamed: 0  \
0  0            
1  1            
2  2            
3  3            
4  4            

                                                                                                                                                                                                                                          tweet  \
0  The only thing I got from college is a caffeine addiction                                                                                                                                                                                      
1  I love it when professors draw a big question mark next to my answer on an exam because I’m always like yeah I don’t either ¯\_(ツ)_/¯                                                                                                          
2  Remember the hundred emails from companies when Covid started getting real? I’ve gotten three in regards to support for protests. And only @SavageXFenty shared helpful links and actually said black lives matter... we love capitalism 🥰🙌🏼   
3  Today my pop-pop told me I was not “forced” to go to college 🙃 okay sure sureeee                                                                                                                                                               
4  @VolphanCarol @littlewhitty @mysticalmanatee I did too, and I also reported Cancun Cruz not worrying about the heartbeats of his constituents without electricity or heat when he fled to Mexico.                                              

   sarcastic  sarcasm  irony  satire  understatement  overstatement  \
0  1          0.0      1.0    0.0     0.0             0.0             
1  1          1.0      0.0    0.0     0.0             0.0             
2  1          0.0      1.0    0.0     0.0             0.0             
3  1          1.0      0.0    0.0     0.0             0.0             
4  1          1.0      0.0    0.0     0.0             0.0             

   rhetorical_question  
0  0.0                  
1  0.0                  
2  0.0                  
3  0.0                  
4  0.0

In [6]:
data = dataset[['tweet', 'sarcastic']]

In [7]:
# create dataset
X = data['tweet'].fillna(' ')
Y = data["sarcastic"]
print(X.shape)
print(Y.shape)



(3468,)
(3468,)


# Feature Extration Using Bag of Words Model

In [8]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer(stop_words='english')
X_bow = vectorizer.fit_transform(X)

In [9]:
analyze = vectorizer.build_analyzer()


In [10]:
vectorizer.get_feature_names()


['00',
 '000',
 '04',
 '06',
 '06z',
 '0and',
 '0c2ifipy8l',
 '0h6fxmjzf5',
 '0qwoj3teph',
 '0zga3r9d1v',
 '10',
 '100',
 '1000',
 '100000',
 '100t',
 '100x',
 '101',
 '104',
 '10pm',
 '10s',
 '10th',
 '11',
 '110',
 '12',
 '121',
 '127',
 '12hrs',
 '12m',
 '13',
 '13th',
 '14',
 '143',
 '14th',
 '15',
 '150',
 '1500',
 '15th',
 '16',
 '165',
 '17',
 '175',
 '18',
 '18th',
 '19',
 '1960s',
 '1966',
 '1989',
 '1995',
 '1997',
 '19th',
 '1ksb5xwxji',
 '1lkhvyupie',
 '1m',
 '1st',
 '1uputyqchg',
 '1vqoutjput',
 '1zfffxpisv',
 '20',
 '200',
 '2000',
 '2000s',
 '2004',
 '2007',
 '200m',
 '2011',
 '2012',
 '2014',
 '2015',
 '2016',
 '2017',
 '2018',
 '2019',
 '2020',
 '2021',
 '2022',
 '2024',
 '2027',
 '206',
 '20s',
 '20th',
 '20years',
 '20yearsofblue',
 '20yr',
 '21',
 '21st',
 '22',
 '22nd',
 '23',
 '2348',
 '24',
 '240',
 '24th',
 '25',
 '256',
 '25pm',
 '26',
 '260k',
 '26th',
 '27',
 '28',
 '29',
 '2am',
 '2br',
 '2days',
 '2e2bcmrovk',
 '2hjxefatmr',
 '2k',
 '2lzu6m7yxx',
 '2nd',
 '

In [11]:
X_bow.toarray()
X_bow.shape

(3468, 10188)

# Feature Extraction Using TfIDF

In [12]:
from sklearn.feature_extraction.text import TfidfTransformer
transformer = TfidfTransformer(smooth_idf=False)
transformer

TfidfTransformer(smooth_idf=False)

In [13]:
from sklearn.feature_extraction.text import TfidfVectorizer

#without smooth IDF
print("Without Smoothing:")
#define tf-idf
tf_idf_vec = TfidfVectorizer(use_idf=True, 
                        smooth_idf=False,  
                        ngram_range=(1,1), stop_words='english') # to use only  bigrams ngram_range=(2,2)
#transform
tf_idf_data = tf_idf_vec.fit_transform(X)
 
#create dataframe
tf_idf_dataframe=pd.DataFrame(tf_idf_data.toarray(),columns = tf_idf_vec.get_feature_names())
print(tf_idf_dataframe)
print("\n")
 
#with smooth

tf_idf_vec_smooth = TfidfVectorizer(use_idf=True,  
                        smooth_idf=True,  
                        ngram_range=(1,1),stop_words='english')
tf_idf_data_smooth = tf_idf_vec_smooth.fit_transform(X)
print("With Smoothing:")
tf_idf_dataframe_smooth = pd.DataFrame(tf_idf_data_smooth.toarray(),columns=tf_idf_vec_smooth.get_feature_names())
print(tf_idf_dataframe_smooth)

Without Smoothing:
       00  000   04   06  06z  0and  0c2ifipy8l  0h6fxmjzf5  0qwoj3teph  \
0     0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          
1     0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          
2     0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          
3     0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          
4     0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          
...   ...  ...  ...  ...  ...  ...   ...         ...         ...          
3463  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          
3464  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          
3465  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          
3466  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          
3467  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0         0.0          

      0zga3r9d1v  ...  zqxelwrbop  zrkxo3w83f  zrtl5alnot  zsl  zsoi0mbkws  \
0 

# Feature Extraction Using Bigrams and Trigrams

In [14]:
c_vec = CountVectorizer(stop_words='english', ngram_range=(2,2))
# matrix of ngrams
ngrams = c_vec.fit_transform(X)
# count frequency of ngrams
count_values = ngrams.toarray().sum(axis=0)
# list of ngrams
vocab = c_vec.vocabulary_
X_2gram  = pd.DataFrame(sorted([(count_values[i],k) for k,i in vocab.items()], reverse=True)
            ).rename(columns={0: 'frequency', 1:'unigram/bigram'})

X_2gram

frequency unigram/bigram
0      25         feel like    
1      17         year old     
2      15         don think    
3      14         ve got       
4      14         just want    
...    ..               ...    
27190  1          000 cases    
27191  1          000 400      
27192  1          00 tip       
27193  1          00 decent    
27194  1          00 24        

[27195 rows x 2 columns]

In [15]:
c_vec = CountVectorizer(stop_words='english', ngram_range=(3,3))
# matrix of ngrams
ngrams = c_vec.fit_transform(X)
# count frequency of ngrams
count_values = ngrams.toarray().sum(axis=0)
# list of ngrams
vocab = c_vec.vocabulary_
X_3gram = pd.DataFrame(sorted([(count_values[i],k) for k,i in vocab.items()], reverse=True)
            ).rename(columns={0: 'frequency', 1:'bigram/trigram'})

X_3gram 

frequency              bigram/trigram
0      4          time office closes        
1      4          team hang 10pm            
2      4          sherlock sherlock sherlock
3      4          service team hang         
4      4          pub crawl pub             
...   ..                    ...             
25295  1          000 cases day             
25296  1          000 400 000               
25297  1          00 tip app                
25298  1          00 decent person          
25299  1          00 24 single              

[25300 rows x 2 columns]

# Binary Classification using TFIDF vectors

In [16]:
#Naïve Bayes, Logistic Regression, Random Forest, SVM, Perceptron

In [17]:
# Machine Learning Models
from sklearn import svm  
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.linear_model import Perceptron
from sklearn.feature_extraction.text import CountVectorizer

In [18]:
# split into train test sets
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.33)
print('Shape of X_train', X_train.shape)
print('Shape of X_test', X_test.shape)
print('Shape of Y_train', y_train.shape)
print('Shape of Y_test', y_test.shape)

Shape of X_train (2323,)
Shape of X_test (1145,)
Shape of Y_train (2323,)
Shape of Y_test (1145,)


In [19]:
max_feature_num = 3468
vectorizer = TfidfVectorizer(max_features=max_feature_num)
train_vecs = vectorizer.fit_transform(X_train)
test_vecs = TfidfVectorizer(max_features=max_feature_num, vocabulary=vectorizer.vocabulary_).fit_transform(X_test)

In [20]:
def NB_classifier(train_vecs, y_train, test_vecs, y_test):
    # Training
    GB = GaussianNB()
    GB.fit(train_vecs, y_train)

    # testing
    test_predictionGB = GB.predict(test_vecs)
    return classification_report(test_predictionGB, y_test) , confusion_matrix(test_predictionGB, y_test)


In [21]:
def LR_classifier(train_vecs, y_train, test_vecs, y_test):
    # Training
    LR = LogisticRegression()
    LR.fit(train_vecs, y_train)

    # testing
    test_predictionLR = LR.predict(test_vecs)
    return classification_report(test_predictionLR, y_test) , confusion_matrix(test_predictionLR, y_test)



In [22]:
def RF_classifier(train_vecs, y_train, test_vecs, y_test):
    # Training
    RF = RandomForestClassifier(n_estimators = 450, max_depth=9, random_state=43)
    RF.fit(train_vecs, y_train)

    # Testing
    test_predictionRF = RF.predict( test_vecs )
    return classification_report(test_predictionRF, y_test), confusion_matrix(test_predictionRF, y_test)

In [23]:
def SVM_classifier(train_vecs, y_train, test_vecs, y_test):
    # Training
    SVM = svm.LinearSVC(max_iter=100)
    SVM.fit(train_vecs, y_train)

    # Testing
    test_predictionSVM = SVM.predict(test_vecs)
    return classification_report(test_predictionSVM, y_test), confusion_matrix(test_predictionSVM, y_test)

In [24]:
def Perceptron_classifier(train_vecs, y_train, test_vecs, y_test):

    # Training
    Per = Perceptron(tol=1e-3, random_state=0)
    Per.fit(train_vecs, y_train)

    # Testing
    test_predictionPer = Per.predict(test_vecs)
    return classification_report(test_predictionPer, y_test), confusion_matrix(test_predictionPer, y_test)


# Results

In [25]:
class_report , conf_matrix = SVM_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of SVM CLASSIFIER on TF-IDF Vectorizer\n')
print(class_report)
print(conf_matrix)


Results of SVM CLASSIFIER on TF-IDF Vectorizer

              precision    recall  f1-score   support

           0       0.89      0.77      0.82       989
           1       0.20      0.37      0.26       156

    accuracy                           0.72      1145
   macro avg       0.55      0.57      0.54      1145
weighted avg       0.79      0.72      0.75      1145

[[763 226]
 [ 98  58]]


In [26]:
class_report , conf_matrix = LR_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of Logistic Regression Classifier on TF-IDF Vectorizer\n')
print(class_report)
print(conf_matrix)

Results of Logistic Regression Classifier on TF-IDF Vectorizer

              precision    recall  f1-score   support

           0       0.99      0.75      0.86      1137
           1       0.01      0.38      0.02         8

    accuracy                           0.75      1145
   macro avg       0.50      0.56      0.44      1145
weighted avg       0.99      0.75      0.85      1145

[[856 281]
 [  5   3]]


In [27]:
class_report , conf_matrix = RF_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of  Random Forest on TF-IDF Vectorizer\n')
print(class_report)
print(conf_matrix)

Results of  Random Forest on TF-IDF Vectorizer

              precision    recall  f1-score   support

           0       1.00      0.75      0.86      1145
           1       0.00      0.00      0.00         0

    accuracy                           0.75      1145
   macro avg       0.50      0.38      0.43      1145
weighted avg       1.00      0.75      0.86      1145

[[861 284]
 [  0   0]]


/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [28]:
class_report , conf_matrix = NB_classifier(train_vecs.todense(), y_train, test_vecs.todense(), y_test)
print('Results of Naive Bayes Classifier on TF-IDF Vectorizer\n')
print(class_report)
print(conf_matrix)

Results of Naive Bayes Classifier on TF-IDF Vectorizer

              precision    recall  f1-score   support

           0       0.55      0.75      0.63       628
           1       0.44      0.24      0.31       517

    accuracy                           0.52      1145
   macro avg       0.49      0.50      0.47      1145
weighted avg       0.50      0.52      0.49      1145

[[470 158]
 [391 126]]


In [29]:
class_report , conf_matrix = Perceptron_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of Perceptron Classifier on TF-IDF Vectorizer\n')
print(class_report)
print(conf_matrix)

Results of Perceptron Classifier on TF-IDF Vectorizer

              precision    recall  f1-score   support

           0       0.76      0.77      0.77       848
           1       0.33      0.31      0.32       297

    accuracy                           0.66      1145
   macro avg       0.55      0.54      0.54      1145
weighted avg       0.65      0.66      0.65      1145

[[657 191]
 [204  93]]


# Binary Classification using BOW

In [30]:
vectorizer = CountVectorizer(binary=False,ngram_range=(1,1))
train_vecs = vectorizer.fit_transform(X_train)
test_vecs = vectorizer.transform(X_test)
print (train_vecs.shape, test_vecs.shape)

(2323, 8021) (1145, 8021)


Note: We will only consider training dataset to define the vocabulary and use the same vocabulary to 
represent the test dataset (as test data is supposed to be hidden).

Thus we will fit our vectorizer on the training data and use it to transform the test data,

There are 8016 unique words in vocabulary.

For each review in our dataset, the Frequency of words(term-frequency) is represented through a vocabulary vector of size 8016. That’s why we have 2323 such vectors in our training-set and similarly 1145 vectors of the similar shape in our test dataset.

Note: binary=False argument means that we fill the vocabulary vector with term-frequency. If binary=True, the vocabulary vector is filled by the presence of words (1 if the word is present and 0 otherwise).



In [31]:
class_report , conf_matrix = SVM_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of SVM CLASSIFIER on BOW\n')
print(class_report)
print(conf_matrix)


Results of SVM CLASSIFIER on BOW

              precision    recall  f1-score   support

           0       0.82      0.78      0.80       916
           1       0.27      0.34      0.30       229

    accuracy                           0.69      1145
   macro avg       0.55      0.56      0.55      1145
weighted avg       0.71      0.69      0.70      1145

[[710 206]
 [151  78]]


/opt/anaconda3/lib/python3.7/site-packages/sklearn/svm/_base.py:986: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [32]:
class_report , conf_matrix = LR_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of Logistic Regression Classifier on BOW \n')
print(class_report)
print(conf_matrix)

Results of Logistic Regression Classifier on BOW 

              precision    recall  f1-score   support

           0       0.91      0.77      0.83      1022
           1       0.16      0.37      0.22       123

    accuracy                           0.72      1145
   macro avg       0.53      0.57      0.53      1145
weighted avg       0.83      0.72      0.77      1145

[[783 239]
 [ 78  45]]


In [33]:
class_report , conf_matrix = RF_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of  Random Forest on BOW \n')
print(class_report)
print(conf_matrix)

Results of  Random Forest on BOW 

              precision    recall  f1-score   support

           0       1.00      0.75      0.86      1145
           1       0.00      0.00      0.00         0

    accuracy                           0.75      1145
   macro avg       0.50      0.38      0.43      1145
weighted avg       1.00      0.75      0.86      1145

[[861 284]
 [  0   0]]


/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [34]:
class_report , conf_matrix = NB_classifier(train_vecs.todense(), y_train, test_vecs.todense(), y_test)
print('Results of Naive Bayes Classifier on BOW \n')
print(class_report)
print(conf_matrix)

Results of Naive Bayes Classifier on BOW 

              precision    recall  f1-score   support

           0       0.68      0.76      0.72       772
           1       0.35      0.27      0.30       373

    accuracy                           0.60      1145
   macro avg       0.52      0.51      0.51      1145
weighted avg       0.58      0.60      0.58      1145

[[588 184]
 [273 100]]


In [35]:
class_report , conf_matrix = Perceptron_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of Perceptron Classifier on BOW\n')
print(class_report)
print(conf_matrix)

Results of Perceptron Classifier on BOW

              precision    recall  f1-score   support

           0       0.82      0.77      0.79       918
           1       0.25      0.32      0.28       227

    accuracy                           0.68      1145
   macro avg       0.54      0.54      0.54      1145
weighted avg       0.71      0.68      0.69      1145

[[706 212]
 [155  72]]


# Binary Classification using Bigram/Trigram

In [36]:
vectorizer = CountVectorizer(binary=False,ngram_range=(2,2))
train_vecs = vectorizer.fit_transform(X_train)
test_vecs = vectorizer.transform(X_test)
print (train_vecs.shape, test_vecs.shape)

(2323, 28181) (1145, 28181)


In [37]:
class_report , conf_matrix = SVM_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of SVM CLASSIFIER on Bi-gram\n')
print(class_report)
print(conf_matrix)

Results of SVM CLASSIFIER on Bi-gram

              precision    recall  f1-score   support

           0       0.96      0.76      0.85      1090
           1       0.08      0.40      0.13        55

    accuracy                           0.74      1145
   macro avg       0.52      0.58      0.49      1145
weighted avg       0.92      0.74      0.81      1145

[[828 262]
 [ 33  22]]


/opt/anaconda3/lib/python3.7/site-packages/sklearn/svm/_base.py:986: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [38]:
class_report , conf_matrix = LR_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of Logistic Regression Classifier on Bi-gram\n')
print(class_report)
print(conf_matrix)

Results of Logistic Regression Classifier on Bi-gram

              precision    recall  f1-score   support

           0       0.99      0.75      0.86      1135
           1       0.01      0.40      0.03        10

    accuracy                           0.75      1145
   macro avg       0.50      0.58      0.44      1145
weighted avg       0.98      0.75      0.85      1145

[[855 280]
 [  6   4]]


In [39]:
class_report , conf_matrix = RF_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of  Random Forest on Bi-gram\n')
print(class_report)
print(conf_matrix)

Results of  Random Forest on Bi-gram

              precision    recall  f1-score   support

           0       1.00      0.75      0.86      1145
           1       0.00      0.00      0.00         0

    accuracy                           0.75      1145
   macro avg       0.50      0.38      0.43      1145
weighted avg       1.00      0.75      0.86      1145

[[861 284]
 [  0   0]]


/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [40]:
class_report , conf_matrix = NB_classifier(train_vecs.todense(), y_train, test_vecs.todense(), y_test)
print('Results of Naive Bayes Classifier on Bi-gram \n')
print(class_report)
print(conf_matrix)

Results of Naive Bayes Classifier on Bi-gram 

              precision    recall  f1-score   support

           0       0.70      0.75      0.72       797
           1       0.30      0.25      0.27       348

    accuracy                           0.60      1145
   macro avg       0.50      0.50      0.50      1145
weighted avg       0.58      0.60      0.59      1145

[[599 198]
 [262  86]]


In [41]:
class_report , conf_matrix = Perceptron_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of Perceptron Classifier on Bi-gram \n')
print(class_report)
print(conf_matrix)

Results of Perceptron Classifier on Bi-gram 

              precision    recall  f1-score   support

           0       0.85      0.76      0.80       959
           1       0.20      0.30      0.24       186

    accuracy                           0.69      1145
   macro avg       0.52      0.53      0.52      1145
weighted avg       0.74      0.69      0.71      1145

[[731 228]
 [130  56]]


In [42]:
vectorizer = CountVectorizer(binary=False,ngram_range=(3,3))
train_vecs = vectorizer.fit_transform(X_train)
test_vecs = vectorizer.transform(X_test)
print (train_vecs.shape, test_vecs.shape)

(2323, 34533) (1145, 34533)


In [43]:
class_report , conf_matrix = SVM_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of SVM CLASSIFIER on Tri-gram\n')
print(class_report)
print(conf_matrix)

Results of SVM CLASSIFIER on Tri-gram

              precision    recall  f1-score   support

           0       1.00      0.75      0.86      1139
           1       0.01      0.50      0.02         6

    accuracy                           0.75      1145
   macro avg       0.50      0.63      0.44      1145
weighted avg       0.99      0.75      0.85      1145

[[858 281]
 [  3   3]]


/opt/anaconda3/lib/python3.7/site-packages/sklearn/svm/_base.py:986: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [44]:
class_report , conf_matrix = LR_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of Logistic Regression Classifier on Tri-gram\n')
print(class_report)
print(conf_matrix)

Results of Logistic Regression Classifier on Tri-gram

              precision    recall  f1-score   support

           0       1.00      0.75      0.86      1143
           1       0.00      0.00      0.00         2

    accuracy                           0.75      1145
   macro avg       0.50      0.38      0.43      1145
weighted avg       1.00      0.75      0.86      1145

[[859 284]
 [  2   0]]


In [45]:
class_report , conf_matrix = RF_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of  Random Forest on Tri-gram\n')
print(class_report)
print(conf_matrix)

Results of  Random Forest on Tri-gram

              precision    recall  f1-score   support

           0       1.00      0.75      0.86      1145
           1       0.00      0.00      0.00         0

    accuracy                           0.75      1145
   macro avg       0.50      0.38      0.43      1145
weighted avg       1.00      0.75      0.86      1145

[[861 284]
 [  0   0]]


/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [46]:
class_report , conf_matrix = NB_classifier(train_vecs.todense(), y_train, test_vecs.todense(), y_test)
print('Results of Naive Bayes Classifier on Tri-gram \n')
print(class_report)
print(conf_matrix)

Results of Naive Bayes Classifier on Tri-gram 

              precision    recall  f1-score   support

           0       0.35      0.79      0.49       383
           1       0.72      0.27      0.39       762

    accuracy                           0.44      1145
   macro avg       0.54      0.53      0.44      1145
weighted avg       0.60      0.44      0.42      1145

[[304  79]
 [557 205]]


In [47]:
class_report , conf_matrix = Perceptron_classifier(train_vecs, y_train, test_vecs, y_test)
print('Results of Perceptron Classifier on Tri-gram \n')
print(class_report)
print(conf_matrix)

Results of Perceptron Classifier on Tri-gram 

              precision    recall  f1-score   support

           0       0.92      0.76      0.84      1039
           1       0.14      0.37      0.20       106

    accuracy                           0.73      1145
   macro avg       0.53      0.57      0.52      1145
weighted avg       0.85      0.73      0.78      1145

[[794 245]
 [ 67  39]]
